pyquil/api/_engagement_manager.py

```diff
@@ -44,28 +45,37 @@ def __init__(self, *, client_configuration: QCSClientConfiguration) -> None:
         :param client_configuration: Client configuration, used for refreshing engagements.
         """
         self._client_configuration = client_configuration
-        self._cached_engagements: Dict[str, EngagementWithCredentials] = {}
+        self._cached_engagements: Dict[Tuple[str, Optional[str]], EngagementWithCredentials] = {}
```

Comment:
Might be better to use a `NamedTuple` or similar to make it more obvious to future devs what these values are.